plugins/net_plugin/net_plugin.cpp

```diff
@@ -315,6 +315,86 @@ namespace eosio {
    constexpr uint32_t signed_block_which       = fc::get_index<net_message, signed_block>();       // see protocol net_message
    constexpr uint32_t packed_transaction_which = fc::get_index<net_message, packed_transaction>(); // see protocol net_message
 
+   class connections_manager {
+      alignas(hardware_destructive_interference_size)
+      mutable std::shared_mutex connections_mtx;
+      std::set<connection_ptr>  connections;
+      chain::flat_set<string>   supplied_peers;
+
+      alignas(hardware_destructive_interference_size)
+      std::mutex                            connector_check_timer_mtx;
+      unique_ptr<boost::asio::steady_timer> connector_check_timer;
+
+      /// thread safe, only modified on startup
+      std::chrono::milliseconds                                heartbeat_timeout{def_keepalive_interval*2};
+      fc::microseconds                                         max_cleanup_time;
+      boost::asio::steady_timer::duration                      connector_period{0};
+      uint32_t                                                 max_client_count{def_max_clients};
+      std::function<void(net_plugin::p2p_connections_metrics)> update_p2p_connection_metrics;
+
+   private: // must call with held mutex
+      connection_ptr find_connection_impl(const string& host) const;
+      void add_impl(connection_ptr&& c);
+      void connect_impl(const string& peer);
+
+      void connection_monitor(const std::weak_ptr<connection>& from_connection);
+
+   public:
+      void add_supplied_peers(const vector<string>& peers ) {
+         std::lock_guard g(connections_mtx);
+         supplied_peers.insert( peers.begin(), peers.end() );
+      }
+
+      // not thread safe, only call on startup
+      void init( std::chrono::milliseconds heartbeat_timeout_ms,
+                fc::microseconds conn_max_cleanup_time,
+                boost::asio::steady_timer::duration conn_period,
+                uint32_t maximum_client_count ) {
+         heartbeat_timeout = heartbeat_timeout_ms;
+         max_cleanup_time = conn_max_cleanup_time;
+         connector_period = conn_period;
+         max_client_count = maximum_client_count;
+      }
+
+      uint32_t get_max_client_count() const { return max_client_count; }
+      fc::microseconds get_connector_period() const {
+         auto connector_period_us = std::chrono::duration_cast<std::chrono::microseconds>( connector_period );
+         return fc::microseconds{ connector_period_us.count() };
+      }
+
+      void register_update_p2p_connection_metrics(std::function<void(net_plugin::p2p_connections_metrics)>&& fun){
+         update_p2p_connection_metrics = std::move(fun);
+      }
+
+      void connect_supplied_peers() {
+         std::lock_guard g(connections_mtx);
+         for (const auto& seed_node : supplied_peers) {
```

Comment:
Why call it `seed_node` and not `peer`?
Since we iterate over `supplied_peers`, it is strange that we do `supplied_peers.insert(host);` in `connect_impl`, as it is likely already there!